Fix multi-site setups active link

alexthn opened the pull request:
On our multi-site setup the active setting isn't working because the paths aren't detected properly. 
We use https://host/welcome for language one and https://host/de/welcome for another.

When you are then on the de/welcome the old code compares de/welcome to welcome which doesn't match. 

This should fix the issue.

alexthn committed:
Fix multi-site setups active link

Fix detection of active parameter on multi-site setups.

engram-design commented:
Let me review this. I did some pretty extensive testing with different multi-site setups not long ago - including using subdomains and sub-directories as the base URL for sites.

Having said that, I believe using `getFullPath()` may not be correct anymore - https://github.com/craftcms/cms/issues/6546

alexthn commented:
I didn't see that one in craft and it does explain. Probably only replacing the getFullPath() with getPathInfo() would also work.

Only thing in that case I wonder about is if you have overlapping URL's on different site's, would that possibly be an issue when not comparing the full url?

engram-design commented:
Probably best-bet to use the absolute URL. I'll run your PR as-is through my tests and see if it's all correct behaviour on my end.

alexthn commented on src/elements/Node.php:
```diff
@@ -600,16 +600,16 @@ public function _getActive($includeChildren = true)
 
         // Stop straight away if this is potentially the homepage
-        if ($currentUrl === '') {
+        if (trim($request->getFullPath(), '/') === '') {
```
```suggestion
        if (trim($request->getPathInfo(), '/') === '') {
```
This could would probably be better with getPathInfo().

engram-design commented:
So reviewing this, there's at least one case this doesn't work for.

If I have two sites, one with `/` as the base URL and another with `/au_en` as the base URL. Then you have a navigation node with a URL of `/au_en/some-page` and this won't pick up the active state correctly, because the `$currentUrl` is now an absolute path instead of relative.

It'll be comparing 'some-page' vs 'au_en/some-page'

Granted, we're using manual links here, but another user had this issue, which prompted my previous thorough investigation.

For your case, the `welcome` page - is this a manual node, or an entry node? If it's an entry node, are you linking to the correct site? For instance, the URL of the "welcome" entry should be `host/welcome` on one site, with `host/de/welcome` on your other site.

This is in reference to your comment:

> When you are then on the de/welcome the old code compares de/welcome to welcome which doesn't match.

`de/welcome` being the `$currentUrl` and `welcome` being `$relativeUrl` - is that correct? Is so, `welcome` doesn't seem like the correct URL for the node...

alexthn commented:
Your right, i didn't think about the manual links because we aren't using them. 
For your your question, all our links are entry's. 


> de/welcome being the $currentUrl and welcome being $relativeUrl - is that correct? Is so, welcome doesn't seem like the correct URL for the node..

This is with an entry and it does seem this is the expected behaviour seeing the craftcms/cms#6546 issue. 

Like stated before it could be an option to just replace the getFullPath() with getPathInfo() its just I wonder if overlapping URL's on different sites could potentially be an issue. 
Alternative solution would be to add a check for this:

```
	// Do the same with the node's URL, get the absolute path, without trailing '/'
        if (strstr($nodeUrl, $siteUrl) === false) {
            // Add the siteUrl since its missing.
            $relativeUrl = $siteUrl.'/'.trim($nodeUrl, '/');
        } else {
            $relativeUrl = trim($nodeUrl, '/');
        }
```

alexthn committed:
Apply suggestions from code review

alexthn committed:
Make manual links work

alexthn commented:
I updated my PR with the suggested changes.
I did some more testing on our site and this seems to be working for us, also with manual links and homepage.

engram-design commented:
Thanks for the adjustments, I'll need to do some further testing, and in some cases the node's URL would be converted to `http://navigation.test/au_en/au_en/manual-page` where there's a duplicate in the siteUrl. 

You might have a manual node with `/au_en/manual-page` and the site URL already uses `/au_en`.

Lots to consider! I'll make some adjustments shortly

engram-design committed:
Cater for nodes that have their siteUrl included in the root-relative node path

- Also renames `relativeUrl` - no longer true. We’re comparing absolute URLs now.
- Also removes some now unneeded code with parent detection.
- Use `UrlHelper::isRootRelativeUrl`

engram-design commented:
I've just made some adjustments which work on my end. Can you check out things on your end?

alexthn commented:
I took a look and looks good to me! Also did some testing, seems perfect!

engram-design commented:
Great, thanks!
